Update Business Details task needs more context

dmallory42 opened the pull request:
Fixes #5069 

#### Changes proposed in this Pull Request

* Update to pass account business errors from Stripe through to the frontend.
* Adds a new modal which will display in the Update Business Details task if more than one error is returned from Stripe.
* Update the task to show the error message in the case of only one error.
* Slightly changes the behaviour depending on whether the account is restricted or restricted soon, to use more clear wording.
* Filter out errors which are vague or unhelpful (currently only one, the `invalid_value_other` code).

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Screenshots

Restricted account with 1 error:

<img width="766" alt="Screenshot 2023-01-18 at 12 01 24" src="https://user-images.githubusercontent.com/11770181/213172062-dc60890a-73d2-4165-90d9-8595f9757026.png">

Restricted account with multiple errors:

<img width="768" alt="Screenshot 2023-01-18 at 12 01 38" src="https://user-images.githubusercontent.com/11770181/213172111-5d2fd46d-a809-4cc9-af2c-a3d67fe6ec15.png">

The New Modal:

<img width="807" alt="Screenshot 2023-01-18 at 12 01 41" src="https://user-images.githubusercontent.com/11770181/213172182-2702c8dc-e7fe-455f-a261-b5cebbd29a73.png">


#### Questions

* Are translations a problem here? We are directly pulling these messages from Stripe. Perhaps we should store the messages on our client to utilise translatable strings? The problem with doing this is that we risk being out of sync with Stripe, if they add or change error messages.

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Check out the server PR (server/2846) and refresh the cache on the client.
* For this PR, we will manually change the data returned from the server to test the different possibilities. I will post the different scenarios below, but for each scenario, replace the array [here](https://github.com/Automattic/woocommerce-payments/blob/114b69231ba44e7e21cf62426620f4fca0b60f60/includes/class-wc-payments-account.php#L191-L206) with the test code, and then ensure the expected behaviour occurs.

**Scenario 1: Restricted Account with Multiple Errors**

* Use the following code:
```php
		return [
			'email'               => $account['email'] ?? '',
			'country'             => $account['country'] ?? 'US',
			'status'              => 'restricted',
			'paymentsEnabled'     => $account['payments_enabled'],
			'deposits'            => $account['deposits'] ?? [],
			'depositsStatus'      => $account['deposits']['status'] ?? $account['deposits_status'] ?? '',
			'currentDeadline'     => $account['current_deadline'] ?? false,
			'pastDue'             => true,
			'accountLink'         => $this->get_login_url(),
			'hasSubmittedVatData' => $account['has_submitted_vat_data'] ?? false,
			'requirements'        => [
				'currentlyDue' => $account['requirements']['currently_due'] ?? [],
				'errors'       => [
										[
						'code'   => 'verification_failed_keyed_identity',
						'reason' => 'The identity information you entered cannot be verified. Please correct any errors or upload a document that matches the identity fields (e.g., name and date of birth) that you entered.',
						'requirement' => 'individual.verification.document',
					],
					[
						'code'   => 'verification_failed_tax_id_match',
						'reason' => 'The tax ID on the account has changed. Please accept the terms of service again.',
						'requirement' => 'tos_acceptance.date',
					],
					[
						'code'   => 'verification_failed_tax_id_match',
						'reason' => 'The tax ID on the account has changed. Please accept the terms of service again.',
						'requirement' => 'tos_acceptance.ip',
					],
					[
						'code' => 'invalid_value_other',
						'reason' => 'This error shouldn\'t display at all',
						'requirement' => 'tos_acceptance.ip',
					],
					[
						'code' => 'invalid_tos_acceptance',
						'reason' => 'This error message should come from our translatable strings',
						'requirement' => 'tos_acceptance.ip',
					],
					[
						'code' => 'new_error_key',
						'reason' => 'This error message should display because we have no matching entry in translatable strings.',
						'requirement' => 'tos_acceptance.ip',
					],
                              ],
			],
		];
```

* The task should be visible on the `Payments/Overview` page, and it should have the `More details` button, which should open a modal with 2 errors displayed (we filter out duplicate error messages).
* Clicking confirm on the modal should open the Stripe Express dashboard. Clicking cancel should close the modal.


**Scenario 2: Restricted Account with Single Error**

```php
		return [
			'email'               => $account['email'] ?? '',
			'country'             => $account['country'] ?? 'US',
			'status'              => 'restricted',
			'paymentsEnabled'     => $account['payments_enabled'],
			'deposits'            => $account['deposits'] ?? [],
			'depositsStatus'      => $account['deposits']['status'] ?? $account['deposits_status'] ?? '',
			'currentDeadline'     => $account['current_deadline'] ?? false,
			'pastDue'             => true,
			'accountLink'         => $this->get_login_url(),
			'hasSubmittedVatData' => $account['has_submitted_vat_data'] ?? false,
			'requirements'        => [
				'currentlyDue' => $account['requirements']['currently_due'] ?? [],
				'errors'       =>  [
					[
						'code'   => 'verification_failed_keyed_identity',
						'reason' => 'The identity information you entered cannot be verified. Please correct any errors or upload a document that matches the identity fields (e.g., name and date of birth) that you entered.',
						'requirement' => 'individual.verification.document',
					],
                              ],
			],
		];
```

* The task should be visible on the `Payments/Overview` page, and it should have the `Update` button. The text should match the error message above.
* The `Update` button should open the Stripe Express Dashboard


**Scenario 3: Restricted Account with No Errors**

```php
		return [
			'email'               => $account['email'] ?? '',
			'country'             => $account['country'] ?? 'US',
			'status'              => 'restricted',
			'paymentsEnabled'     => $account['payments_enabled'],
			'deposits'            => $account['deposits'] ?? [],
			'depositsStatus'      => $account['deposits']['status'] ?? $account['deposits_status'] ?? '',
			'currentDeadline'     => $account['current_deadline'] ?? false,
			'pastDue'             => true,
			'accountLink'         => $this->get_login_url(),
			'hasSubmittedVatData' => $account['has_submitted_vat_data'] ?? false,
			'requirements'        => [
				'currentlyDue' => $account['requirements']['currently_due'] ?? [],
				'errors'       => $account['requirements']['errors'] ?? [],
			],
		];
```

* The task should be visible on the `Payments/Overview` page, and it should have the `Update` button. The text should be a generic message about updating business details.
* The `Update` button should open the Stripe Express Dashboard

**Scenario 4: Restricted Soon Account with Multiple Errors**

```php
		return [
			'email'               => $account['email'] ?? '',
			'country'             => $account['country'] ?? 'US',
			'status'              => 'restricted',
			'paymentsEnabled'     => $account['payments_enabled'],
			'deposits'            => $account['deposits'] ?? [],
			'depositsStatus'      => $account['deposits']['status'] ?? $account['deposits_status'] ?? '',
			'currentDeadline'     => 1674648153,
			'pastDue'             => false,
			'accountLink'         => $this->get_login_url(),
			'hasSubmittedVatData' => $account['has_submitted_vat_data'] ?? false,
			'requirements'        => [
				'currentlyDue' => $account['requirements']['currently_due'] ?? [],
				'errors'       => [
						[
						'code'   => 'verification_failed_keyed_identity',
						'reason' => 'The identity information you entered cannot be verified. Please correct any errors or upload a document that matches the identity fields (e.g., name and date of birth) that you entered.',
						'requirement' => 'individual.verification.document',
					],
					[
						'code'   => 'verification_failed_tax_id_match',
						'reason' => 'The tax ID on the account has changed. Please accept the terms of service again.',
						'requirement' => 'tos_acceptance.date',
					],
					[
						'code'   => 'verification_failed_tax_id_match',
						'reason' => 'The tax ID on the account has changed. Please accept the terms of service again.',
						'requirement' => 'tos_acceptance.ip',
					],
					[
						'code' => 'invalid_value_other',
						'reason' => 'This error shouldn\'t display at all',
						'requirement' => 'tos_acceptance.ip',
					],
					[
						'code' => 'invalid_tos_acceptance',
						'reason' => 'This error message should come from our translatable strings',
						'requirement' => 'tos_acceptance.ip',
					],
					[
						'code' => 'new_error_key',
						'reason' => 'This error message should display because we have no matching entry in translatable strings.',
						'requirement' => 'tos_acceptance.ip',
					],
                              ],
			],
		];
```

* The task should be visible on the `Payments/Overview` page, and it should have the `More details` button, which should open a modal with 2 errors displayed (we filter out duplicate error messages). It should also show the date to update by.
* Clicking confirm on the modal should open the Stripe Express dashboard. Clicking cancel should close the modal.


**Scenario 5: Restricted Soon Account with Single Error**

```php
		return [
			'email'               => $account['email'] ?? '',
			'country'             => $account['country'] ?? 'US',
			'status'              => 'restricted',
			'paymentsEnabled'     => $account['payments_enabled'],
			'deposits'            => $account['deposits'] ?? [],
			'depositsStatus'      => $account['deposits']['status'] ?? $account['deposits_status'] ?? '',
			'currentDeadline'     => $account['current_deadline'] ?? false,
			'pastDue'             => false,
			'accountLink'         => $this->get_login_url(),
			'hasSubmittedVatData' => $account['has_submitted_vat_data'] ?? false,
			'requirements'        => [
				'currentlyDue' => $account['requirements']['currently_due'] ?? [],
				'errors'       => [
					[
						'code'   => 'verification_failed_keyed_identity',
						'reason' => 'The identity information you entered cannot be verified. Please correct any errors or upload a document that matches the identity fields (e.g., name and date of birth) that you entered.',
						'requirement' => 'individual.verification.document',
					],
                              ],
			],
		];
```

* The task should be visible on the `Payments/Overview` page, and it should have the `Update` button. The text should match the error message above and also have the date needed to update by.
* The `Update` button should open the Stripe Express Dashboard

**Scenario 6: Restricted Soon Account with No Errors**

```php
		return [
			'email'               => $account['email'] ?? '',
			'country'             => $account['country'] ?? 'US',
			'status'              => 'restricted_soon',
			'paymentsEnabled'     => $account['payments_enabled'],
			'deposits'            => $account['deposits'] ?? [],
			'depositsStatus'      => $account['deposits']['status'] ?? $account['deposits_status'] ?? '',
			'currentDeadline'     => 1674648153,
			'pastDue'             => true,
			'accountLink'         => $this->get_login_url(),
			'hasSubmittedVatData' => $account['has_submitted_vat_data'] ?? false,
			'requirements'        => [
				'currentlyDue' => $account['requirements']['currently_due'] ?? [],
				'errors'       => $account['requirements']['errors'] ?? [],
			],
		];
```

* The task should be visible on the `Payments/Overview` page, and it should have the `Update` button. The text should be a generic message about updating business details with the date to update by.
* The `Update` button should open the Stripe Express Dashboard

**Scenario 7: Account in Good Standing**

* Use a normal, active Stripe account which is not restricted. 
* Use the default code, shared below: 
```php
		return [
			'email'               => $account['email'] ?? '',
			'country'             => $account['country'] ?? 'US',
			'status'              => $account['status'],
			'paymentsEnabled'     => $account['payments_enabled'],
			'deposits'            => $account['deposits'] ?? [],
			'depositsStatus'      => $account['deposits']['status'] ?? $account['deposits_status'] ?? '',
			'currentDeadline'     => $account['current_deadline'] ?? false,
			'pastDue'             => true,
			'accountLink'         => $this->get_login_url(),
			'hasSubmittedVatData' => $account['has_submitted_vat_data'] ?? false,
			'requirements'        => [
				'currentlyDue' => $account['requirements']['currently_due'] ?? [],
				'errors'       => $account['requirements']['errors'] ?? [],
			],
		];
```
* The task should not appear.


-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [x] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [x] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

dmallory42 committed:
Updates to the format of the wcpaySettings interface, and update to pass the requirements from the server to the client JS.

dmallory42 committed:
WIP: create the business details modal and set up test data

dmallory42 committed:
Merge remote-tracking branch 'origin/trunk' into add/5069-update-business-details-task

dmallory42 committed:
Merge remote-tracking branch 'origin/develop' into add/5069-update-business-details-task

dmallory42 committed:
Updates to support showing the modal in cases where there is more than one error.

dmallory42 committed:
Updates to the modal logic and the task list.

dmallory42 committed:
Updates to the modal styling and logic to display it.

dmallory42 committed:
Update changelog

dmallory42 committed:
Merge branch 'develop' into add/5069-update-business-details-task

dmallory42 commented on includes/class-wc-payments-account.php:
```diff
@@ -196,9 +196,13 @@ public function get_account_status_data() {
 			'depositsStatus'      => $account['deposits']['status'] ?? $account['deposits_status'] ?? '',
 			'currentDeadline'     => $account['current_deadline'] ?? false,
-			'pastDue'             => $account['has_overdue_requirements'] ?? false,
+			'pastDue'             => true,
```
This is just for testing, will remove before final review.

ismaeldcom commented on client/overview/task-list/tasks.js:
```diff
@@ -20,15 +23,57 @@ export const getTasks = ( {
+		);
+	};
+
+	const getErrorMessagesFromRequirements = ( requirements ) => {
+		if ( 'errors' in requirements ) {
+			const errorMessages = requirements.errors.map( ( error ) => {
+				return error.reason;
+			} );
+			return [ ...new Set( errorMessages ) ];
+		}
+
+		return [];
+	};
```
When we define a pure function like this one, we can move it outside the exported one, or even in a different file, if that help us organize better the code 🙂 

As a bonus tip, we can rewrite it using new ES features to make it shorter and hopefully more readable:
```js
	const getErrorMessagesFromRequirements = ( requirements ) => [
		...new Set( requirements?.errors?.map( ( error ) => error.reason ) ),
	];
```

Although, as you already provided a fallback value for `requirements.errors` in the PHP side: https://github.com/Automattic/woocommerce-payments/blob/114b69231ba44e7e21cf62426620f4fca0b60f60/includes/class-wc-payments-account.php#L204

We can get rid of `?.` and go with just:
```js
	const getErrorMessagesFromRequirements = ( requirements ) => [
		...new Set( requirements.errors.map( ( error ) => error.reason ) ),
	];
```

dmallory42 replied:
Thanks @ismaeldcom, that's a great tip! I updated in the latest commits, but I kept the `?` operators because I found if the account is not fetched properly (e.g. we can't fetch it from the server) then the page will break if we don't check for the parameters. (normally in this case we'd see the `Error determining account status` fallback).

ismaeldcom commented on client/globals.d.ts:
```diff
@@ -10,10 +16,21 @@ declare const wcpaySettings: {
+		pastDue?: boolean;
+		accountLink: string;
 		hasSubmittedVatData?: boolean;
+		requirements?: {
+			currentlyDue?: Array< string >;
+			errors?: Array< RequirementError >;
+		};
```
As we return fallback values for those properties from the PHP side. We should make them non-optional by removing the `?`.

ismaeldcom commented on client/overview/task-list/tasks.js:
```diff
@@ -20,15 +23,57 @@ export const getTasks = ( {
+		status,
+		accountLink,
+		currentDeadline
+	) => {
+		const container = document.createElement( 'div' );
+		container.id = 'wcpay-update-business-details-container';
+		document.body.appendChild( container );
+		ReactDOM.render(
+			<UpdateBusinessDetailsModal
```
We're doing this in a really hacky way 😅 As we don't have an identifiable element, and an event to trigger the component mount. I think your approach of rendering it on task click event is okay. Although we could not show the modal on `More details` click, like in the design.

The current implementation will keep adding new containers every time we close the modal and click on the task. To fix that we need to add it only once, and provide a `key` to the component so the state get discard, and it does not remain closed. Something like:
```js
		let container = document.querySelector(
			'#wcpay-update-business-details-container'
		);
		if ( ! container ) {
			container = document.createElement( 'div' );
			container.id = 'wcpay-update-business-details-container';
			document.body.appendChild( container );
		}
		ReactDOM.render(
			<UpdateBusinessDetailsModal
				key={ Date.now() }
```

dmallory42 replied:
Yeah - it's definitely not ideal 😅 thanks for the tip on not injecting a new container each time.

dmallory42 committed:
Fixes based on initial review to make code more efficient and performant.

dmallory42 committed:
Merge branch 'add/5069-update-business-details-task' of github.com:Automattic/woocommerce-payments into add/5069-update-business-details-task

dmallory42 committed:
Fixes based on initial review to make code more efficient and performant.

dmallory42 committed:
Added snapshot tests for the new modal.

dmallory42 committed:
Added logic to filter out error messages which are not helpful.

dmallory42 committed:
Adding comment to give more context.

dmallory42 commented:
Hi @elizaan36 - would it be possible to get a design 👀  on this to make sure it all looks as expected? Let me know if you have any questions or need any more screenshots/demos. Thanks!

ismaeldcom commented:
Thanks for the detailed PR description and instructions. With the latest changes everything works as expected. 🎉 

I think it's a good compromise if we consider our current limitations.

Only one nitpick left 🙂  for  `Scenario 5: Restricted Soon Account with Single Error` I think we should move the `Update by %s to avoid a disruption in deposits.` part to a new line. To make it easier to read and differentiate from the reason.

>Are translations a problem here? We are directly pulling these messages from Stripe. Perhaps we should store the messages on our client to utilise translatable strings? The problem with doing this is that we risk being out of sync with Stripe, if they add or change error messages.

I missed it before. Good question! I personally think we should try to internationalize whenever we can. For this case, we already did something similar with [intent error messages](https://github.com/Automattic/woocommerce-payments/blob/trunk/includes/class-wc-payments-webhook-processing-service.php#L712).

Being out of sync is quite possible, but falling back to Stripe's message will ensure we at least provide an English version in those cases.

elizaan36 commented:
Nice @dmallory42 ! 🎯

A few questions/notes:
- The task should automatically dismiss once the account is in good standing again, is that possible?
- Maybe out of scope for this particular issue (because I have a hunch this is related to the WordPress components package), but the `Things to do` header is misaligned.

<img width="160" alt="image" src="https://user-images.githubusercontent.com/5121465/213504356-1b7d3021-89c5-4456-953d-5cac5bb38cb5.png">

The title should be the larger `Title` text size, and aligned horizontally with the count badge.

- Lastly, the notice copy in the modal should be horizontally aligned within the notice background. So the padding is equal top and bottom.

<img width="324" alt="image" src="https://user-images.githubusercontent.com/5121465/213505287-22a988d3-c3a4-4011-b357-e3af805b1631.png">

Everything else looks great! I can only view the screenshots so far so if you have a minute to share a quick video of the click through that would be great, otherwise I trust that you've tested.

dmallory42 committed:
Merge branch 'develop' into add/5069-update-business-details-task

dmallory42 commented:
Thanks for looking, @elizaan36!

> The task should automatically dismiss once the account is in good standing again, is that possible?

Yep, this is the current behaviour :) this task would only show if an account is in a `restricted` or `restricted_soon` state.

>  Maybe out of scope for this particular issue (because I have a hunch this is related to the WordPress components package), but the Things to do header is misaligned. The title should be the larger Title text size, and aligned horizontally with the count badge.

Good catch! This is actually related to how we're using the components in WCPay, so we can fix it. I chatted with @ismaeldcom about it on Slack and I'm going to tackle it in a new issue to avoid doing too much in one issue.

> Lastly, the notice copy in the modal should be horizontally aligned within the notice background. So the padding is equal top and bottom.

Good catch, thanks! Will fix this before going ahead, and can also create a few GIFs to show the final look/behaviour.

dmallory42 committed:
Various fixes to the logic, adding i8n strings and adding a new line to show the date more clearly to the user.

dmallory42 committed:
Merge branch 'add/5069-update-business-details-task' of github.com:Automattic/woocommerce-payments into add/5069-update-business-details-task

dmallory42 committed:
Fixing the padding on the notices, update snapshots.

dmallory42 committed:
Fix broken test.

dmallory42 commented:
@elizaan36 sharing some GIFs of the functionality, the messaging is slightly different based on if it is a restricted or a restricted soon account, but I hope this gives an idea. 🙂 Also, the error messages are just test/examples - the actual messages will match the ones provided by Stripe for a particular account.

![Screen Capture on 2023-01-20 at 15-00-26](https://user-images.githubusercontent.com/11770181/213730745-b04cbf67-b979-4c58-8822-a91811e3b1f1.gif)

@ismaeldcom  - I've now added the translatable strings and updated the logic to filter out certain error messages. Could you take a final quick look please? I slightly updated the test instructions so that the examples test the new functionality, hope it all makes sense!

elizaan36 commented:
Many thanks @dmallory42 🙏

Looks good. The only thing I'd change is in the modal, the "Additional information..." and "Update by..." sentences could be in the same line/paragraph for smoother readability. I don't think the line break is necessary since the timeframe is also provided in the task.

dmallory42 committed:
Merge branch 'develop' into add/5069-update-business-details-task

ismaeldcom reviewed (approved):
Thanks for making those changes @dmallory42. The internationalization works as expected, and now we can even add links in the reason, as you did! 🎉  There is no issue following testing instructions and the code looks good.

Preapproving it now! The only thing left is the tiny detail from @elizaan36

dmallory42 committed:
Merge branch 'develop' into add/5069-update-business-details-task

dmallory42 committed:
Update to remove linebreak in the description within the modal.
